### PR TITLE
Skip loading .data bundle and mapping corresponding files when running as a worker for pthread

### PR DESCRIPTION
**Pull request by shanumantesc**
Addresses https://github.com/emscripten-core/emscripten/issues/15627

### CODE REVIEW

**Comment by shanumantesc** [comment]
cc @sbc100 

What would be a good way to test this; or is the file-system functionality already covered by existing unit tests?

File: tools/file_packager.py
Comment (sbc100) [inline tools/file_packager.py:295]: I'm not sure what you are trying do here.. it looks like you are setting this to the empty list if and only if its already empty.

Also, `preRun` callbacks are only ever run once on the main thread.     See here where we call preRun there is an early return if we are running on a thread so we never get here:  https://github.com/emscripten-core/emscripten/blob/ae5c256cacb32b1a0f303566ee7ac18448a9d7fc/src/postamble.js#L280

Comment (shanumantesc) [reply tools/file_packager.py:295]: Hmm, I added this logic because the generated glue file was calling Array.slice on `Module['preRun']` (and so used the code from https://github.com/emscripten-core/emscripten/blob/main/tools/file_packager.py#L876 ), 

But I no longer see it with version 3.0.0. Maybe something changed around his, or I had some local changes to my toolchain. Will revert.

File: tools/file_packager.py
Comment (sbc100) [inline tools/file_packager.py:294]: Looking at the code it looks like we actually run the code here by adding it to `preRun` or by calling runWithFS directly in the case that `Module['calledRun']` is set.

Presumably the issue you are seeing must be because Module['calledRun'] is set in the pthreads?

```
    if (Module['calledRun']) {    
      runWithFS();    
    } else {    
      if (!Module['preRun']) Module['preRun'] = [];                                               
      Module["preRun"].push(runWithFS); // FS is not initialized yet, wait for it                 
    }  
```

@kripken I think you have a better understanding of this code than I do.

Whatever we do we should probably try to write a test for this..

Comment (shanumantesc) [reply tools/file_packager.py:294]: The overall structure I see is
```
(function() {
  var loadPackage = function() {
    fetchRemotePackage = function(...) {}; // responsible for downloading .data 
    fetchRemotePackage(...); // always called
    runWithFS = function() {...}; // maps downloaded .data to file-system paths
    if (Module['calledRun']) { ... } else { ... } // Call runWithFS() 
  } 
  loadPackage(file_mappings); 
}(); <--- Currently always called on pthread worker
```

The issue is that since the above closure is always called on the worker, we always call `fetchRemotePackage` which will make a request to download the .data file. On worker, this .data file is always assumed to be at the server root, and `getPreloadedPackage` is ignored (as mentioned in the linked issue) which results in a failure to download .data file.

The initial solution I had was to build a mechanism to provide a correctly overloaded `getPreloadedPackage` to the worker and still allow runWithFS() to execute. However, since file-system calls will be proxied from pthread anyways, we can skip the entire `loadPackage` function (per my understanding). Hence, I've modified loadPacakage to early return in case we're running on a pthread worker.

Re: unit test, I was looking to add something, but didn't find any closely matching test. Would be happy to get pointers on how we can test this if not covered by existing tests.

Comment (sbc100) [reply tools/file_packager.py:294]: I think the convention is to not include { } around the single-line block

**Review by sbc100 — COMMENTED** [review]
In general this lgtm, but I'm not very familiar with this code so looping in @kripken

**Comment by sbc100** [comment]
As for testing I would take a look at the tests in test_other.py that use `--preload-file`.   I imagine you want to create a new test called something like `test_pthread_preload_files`.

**Comment by shanumantesc** [comment]
Nudge for this one @kripken

**Comment by shanumantesc** [comment]
@sbc100 @kripken This the final piece in the puzzle to get a threaded build working for our use case, I would really appreciate if you could go over this change to help get it into the next emscripten release :) 
Thanks!

File: tools/file_packager.py
Comment (sbc100) [inline tools/file_packager.py:300]: Indentation here should be 2 spaces.

File: tools/file_packager.py
Comment (sbc100) [inline tools/file_packager.py:299]: This doesn't seem like it should be necessary.. we should probably fix js_manipulation.py to be smarter... I'll honestly don't know what it that is for but I will check it out.

File: tools/file_packager.py
Comment (sbc100) [inline tools/file_packager.py:296]: Are we sure that `ENVIRONMENT_IS_PTHREAD` is always exported on the Module object in pthreads mode?

File: tools/file_packager.py
Comment (sbc100) [reply tools/file_packager.py:299]: It looks like we can modify `js_manipulation.py` such that it also wraps its logic in `if (Module['ENVIRONMENT_IS_PTHREAD']) {`

File: tools/file_packager.py
Comment (shanumantesc) [reply tools/file_packager.py:296]: It should be guaranteed, since we rely on it further down to set `ENVIRONMENT_IS_PTHREAD` 
https://github.com/emscripten-core/emscripten/blob/0ca340670908acb5ac917fa123ce24c6443bca56/src/shell.js#L125

File: tools/file_packager.py
Comment (shanumantesc) [reply tools/file_packager.py:299]: Thanks, let me know if there is a better way to fix this. 
It looks like with assertions enabled, we are trying to enforce that the .data file has been processed (https://github.com/emscripten-core/emscripten/blob/4d864df0a57024d6/tools/js_manipulation.py#L22) and the presence of `Module['preRun']` is used to determine that.

File: tools/file_packager.py
Comment (shanumantesc) [reply tools/file_packager.py]: fixed

File: tests/test_other.py
Comment (shanumantesc) [reply tests/test_other.py:1491]: Yes - if I only change file_packager.py, but don't change js_manipulation.py, then the test which proxies to pthread fails with error
```
worker.js onmessage() captured an uncaught exception: TypeError: Cannot read property 'slice' of undefined
pthread sent an error! undefined:undefined: Cannot read property 'slice' of undefined

/private/var/folders/70/j6dv6jq15ml8729ps84cctl00000gp/T/tmppl3nk87t/emscripten_test_other_x02xzu0e/a.out.js:360
      throw ex;
```

The default test is to ensure that changes to `file_packager.py` is not breaking preloading with pthread enabled

Comment (sbc100) [reply tests/test_other.py:1491]: Sorry, I mean if you revert everything except this test.. does it pass?   Another way of putting it:  Does this test fail without the fixes in this PR?   If not, can we make it so that it does?

Comment (sbc100) [reply tests/test_other.py:1491]: We would need to be able to detect the fact that the files were re-loaded when a thread was first created I guess?   How did you observe this bug originally?  i.e. what was the side effect?

Comment (shanumantesc) [reply tests/test_other.py:1491]: > Does this test fail without the fixes in this PR? If not, can we make it so that it does?

I couldn't find the right way to do this -- I observed the failure when our .data file is not hosted in the server root (and so is provided via overloading getPreloadedPackage). In this case, making a request for `/MyLibrary.data` fails. Can you point me to how we can make it so that the .data file is not available by default in server root and requires use of `getPreloadedPackage`? 

Overall, you are right that the tests don't explicitly check what was failing, but rather are ensuring that early return isn't breaking file preloading.

Comment (shanumantesc) [reply tests/test_other.py:1491]: nvm, found the test in test_browser.py; will add another one

**Comment by shanumantesc** [comment]
Added a test which fails without changes in this PR with error
```
[client logging: /?exception=Uncaught [object ErrorEvent] / undefined ]
[client logging: /?exception=Uncaught Error: File not found : http://localhost:8888/manual_download_data.data / undefined ]
FAIL
```

File: tests/test_other.py
Comment (shanumantesc) [reply tests/test_other.py:1491]: Should I remove the `test_preload_with_pthread` tests, or are we okay to keep both the browser test and this one @sbc100 ?

File: tests/manual_download_data_pthread.html
Comment (sbc100) [inline tests/manual_download_data_pthread.html:194]: Adding this entire file makes it hard to see what is being tested.

Can we not just take this two JS functions and put them a in `--pre-js` files?   

Also, would it be enough to just do this in a pre-js?

```
Module.getPreloadedPackage = function(remotePackageName, remotePackageSize) {
  console.log('getPreloadedPackage called')
  assert(!ENVIRONMENT_IS_PHTREAD, "getPreloadedPackage should only ever be called from the main thread");
  // We can't/don't actually do any pre-loading, we are just checked that we are never called from a thread.
  return false;
}
```

I think it we can simplify this test to just this we don't need to also add a test to test_core.py.

Comment (shanumantesc) [reply tests/manual_download_data_pthread.html:194]: Overloading `Module.getPreloadedPackage` will still fail because when the generated worker file loads the main glue file, it doesn't set `Module.getPreloadedPackage`. That is why the glue file tries to access  `/MyLibrary.data` from server root. Modifying worker.js to set `Module.getPreloadedPackage` was a part of my very first proposed solution/commit, but it is unnecessary given that file system is being proxied.

Re: just using `--pre-js` files, I'm not sure if I can use that to test my exact error before these changes - which is that the preload phase fails when the .data file is not at server root, while I we are replicating that exact scenario in this test case. Do you see a way to test it accurately with just pre-js files?

Comment (shanumantesc) [reply tests/manual_download_data_pthread.html:194]: Another point in favor of current test is that the existing `test_preload_file_with_manual_data_download` was never checking for pthreads, which is why the bug wasn't caught before. So IMO we should have an equivalent test with pthreads enabled. I can combine the two tests to use the same .html file instead of creating a new `tests/manual_download_data_pthread.html` if that helps reduce the noise.

File: tests/test_browser.py
Comment (sbc100) [inline tests/test_browser.py:418]: Maybe needs `@requires_threads` on this test now?

File: tests/manual_download_data.html
Comment (sbc100) [inline tests/manual_download_data.html:192]: What does this change do?

File: tests/manual_download_data.html
Comment (sbc100) [inline tests/manual_download_data.html:189]: Why remove this?

**Review by sbc100 — COMMENTED** [review]
lgtm.  Just a couple more questions.

Sorry for all the back and forth.. i think we are almost ready to land this.

File: tests/manual_download_data.html
Comment (shanumantesc) [reply tests/manual_download_data.html:192]: It is used by the worker to load the glue file - https://github.com/emscripten-core/emscripten/blob/df034c938c8bdbb3b0a272e9794084e9bce8567c/src/library_pthread.js#L330

Comment (sbc100) [reply tests/manual_download_data.html:192]: > It is used by the worker to load the glue file -
> 
> https://github.com/emscripten-core/emscripten/blob/df034c938c8bdbb3b0a272e9794084e9bce8567c/src/library_pthread.js#L330

So it wasn't needed before because this test didn't run with workers/threads?

File: tests/manual_download_data.html
Comment (shanumantesc) [reply tests/manual_download_data.html:189]: This check is redundant - here we check if `getPreloadedPackage` method in JS was called by checking for `Module.manuallyDownloadedData` ; however, we are already checking for the file existence and it's contents, which can only succeed if `getPreloadedPackage` was called (since .data is not available any more in server root)

File: tests/manual_download_data.html
Comment (sbc100) [reply tests/manual_download_data.html:192]: If you don't do this will the test not pass?

File: tests/test_browser.py
Comment (shanumantesc) [reply tests/test_browser.py:418]: Is it required even when using `self.run_browser` ?

File: tests/manual_download_data.html
Comment (sbc100) [reply tests/manual_download_data.html:189]: What if `getPreloadedPackage` didn't exist or didn't run though (e.g. if it was renamed)?  Wouldn't the file_packager code still run to load the preloaded data in the normal way that `--preload-file` does?   .. and therefore the file would exist even if `getPreloadedPackage` wasn't called?

File: tests/test_browser.py
Comment (sbc100) [reply tests/test_browser.py:418]: Yes, that idea is that its possible to run the entire test suite in a browser that doesn't support threads, in which case such tests as these would need to be skipped.

File: tests/manual_download_data.html
Comment (shanumantesc) [reply tests/manual_download_data.html:192]: >  So it wasn't needed before because this test didn't run with workers/threads?
yes

> If you don't do this will the test not pass?

yes, pthread test won't pass

File: tests/manual_download_data.html
Comment (shanumantesc) [reply tests/manual_download_data.html:189]: > Wouldn't the file_packager code still run to load the preloaded data in the normal way that --preload-file does?

Yes. But it would fail to load the file the normal way (which is to try and find the .data file at the root of the server)

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
lgtm with passing test.  Would still like to get a second opinion from @kripken

**Comment by shanumantesc** [comment]
Please take a look @kripken , hoping to get this change upstream soon to unblock pthread support for our use case

**Pull request merged**
